hack/dockerfiles/test.buildkit.Dockerfile

```diff
@@ -0,0 +1,236 @@
+
+FROM registry:$REGISTRY_VERSION AS registry
+
+FROM gobuild-base AS rootlesskit
+ARG ROOTLESSKIT_VERSION
+RUN git clone https://github.com/rootless-containers/rootlesskit.git /go/src/github.com/rootless-containers/rootlesskit
+WORKDIR /go/src/github.com/rootless-containers/rootlesskit
+ARG TARGETPLATFORM
+RUN  --mount=target=/root/.cache,type=cache \
+  git checkout -q "$ROOTLESSKIT_VERSION"  && \
+  CGO_ENABLED=0 go build -o /rootlesskit ./cmd/rootlesskit && \
+  file /rootlesskit | grep "statically linked"
+
+# Copy together all binaries needed for oci worker mode
+FROM buildkit-export AS buildkit-buildkitd.oci_only
+COPY --from=buildkitd.oci_only /usr/bin/buildkitd.oci_only /usr/bin/
+COPY --from=buildctl /usr/bin/buildctl /usr/bin/
+ENTRYPOINT ["buildkitd.oci_only"]
+
+# Copy together all binaries for containerd worker mode
+FROM buildkit-export AS buildkit-buildkitd.containerd_only
+COPY --from=buildkitd.containerd_only /usr/bin/buildkitd.containerd_only /usr/bin/
+COPY --from=buildctl /usr/bin/buildctl /usr/bin/
+ENTRYPOINT ["buildkitd.containerd_only"]
+
+# Copy together all binaries for oci+containerd mode
+FROM buildkit-export AS buildkit-buildkitd
+COPY --from=binaries / /usr/bin/
+ENTRYPOINT ["buildkitd"]
+
+FROM alpine AS containerd-runtime
+COPY --from=runc /usr/bin/runc /usr/bin/
+COPY --from=containerd /out/containerd* /usr/bin/
+COPY --from=containerd /out/ctr /usr/bin/
+VOLUME /var/lib/containerd
+VOLUME /run/containerd
+ENTRYPOINT ["containerd"]
+
+FROM buildkit-base AS integration-tests
+ENV BUILDKIT_INTEGRATION_ROOTLESS_IDPAIR="1000:1000"
+RUN apt-get install -y --no-install-recommends uidmap sudo \
+  && useradd --create-home --home-dir /home/user --uid 1000 -s /bin/sh user \
+  && echo "XDG_RUNTIME_DIR=/run/user/1000; export XDG_RUNTIME_DIR" >> /home/user/.profile \
+  && mkdir -m 0700 -p /run/user/1000 \
+  && chown -R user /run/user/1000 /home/user
+ENV BUILDKIT_INTEGRATION_CONTAINERD_EXTRA="containerd-1.0=/opt/containerd-1.0/bin"
+COPY --from=rootlesskit /rootlesskit /usr/bin/
+COPY --from=containerd10 /out/containerd* /opt/containerd-1.0/bin/
+COPY --from=registry /bin/registry /usr/bin
+COPY --from=containerd /out/containerd* /usr/bin/
+COPY --from=binaries / /usr/bin/
+COPY . .
+
+FROM rootless
+ENV BUILDKIT_INTEGRATION_ROOTLESS_IDPAIR="1000:1000"
+RUN apt-get install -y uidmap sudo \
+  && useradd --create-home --home-dir /home/user --uid 1000 -s /bin/sh user \
+  && echo "XDG_RUNTIME_DIR=/run/user/1000; export XDG_RUNTIME_DIR" >> /home/user/.profile \
+  && mkdir -m 0700 -p /run/user/1000 \
+  && chown -R user /run/user/1000 /home/user
+
+FROM alpine AS rootless-base
+RUN apk add --no-cache shadow shadow-uidmap \
+  && useradd --create-home --home-dir /home/user --uid 1000 user \
+  && mkdir -p /run/user/1000 /home/user/.local/tmp /home/user/.local/share/buildkit \
+  && chown -R user /run/user/1000 /home/user
+
+# Rootless mode.
+# Still requires `--privileged`.
+FROM tonistiigi/buildkit:rootless-base AS rootless
```

Comment:
btw, rootlesskit multi-platform is currently blocked with https://github.com/moby/datakit/pull/638

Reply:
wow, didn't notice rootlesskit was vendoring datakit

Reply:
why do we need the `copy` binary for `rootless`?

Reply:
I just used to `copy` Dockerfile as an example of the same technique. How the `BASEMODE` arg there is used to switch between already built base image and a stage. The `copy` binary is not in the `rootless-base` image,  it is the same stage as in this file: alpine with shadow packages and `user` setup.

Reply:
sorry I'm still confused - why can't use the direct `rootless-base` defined in this Dockerfile?
Doesn't `FROM --platform=$BUILDPLATFORM alpine AS rootless-base` work?

Reply:
Because then a node that only supports execution in `amd64` (like Travis node) can't build the rootless target for all platforms(arm, ppc64le etc) because it can't run `apk/useradd` for these architectures. If I'd add `--platform=$BUILDPLATFORM` that would always use amd64 packages on travis and install amd64 packages making the resulting images not multi-platform.

Reply:
Does `FROM --platform=$TARGETPLATFORM alpine AS rootless-base` work?

Reply:
Only if node can run `apk` an all platforms. What travis can't. The difference is that `tonistiigi/buildkit:rootless-base` is already multi-platform so any node can add the missing binaries on top of that.

Also, `FROM --platform=$TARGETPLATFORM alpine AS rootless-base` is the same as `FROM alpine AS rootless-base`.

```
 # docker run --rm mplatform/mquery tonistiigi/buildkit:rootless-base
Image: tonistiigi/buildkit:rootless-base
 * Manifest List: Yes
 * Supported platforms:
   - linux/amd64
   - linux/arm/v6
   - linux/arm/v7
   - linux/arm64/undefined
   - linux/ppc64le
   - linux/s390x
```

Reply:
> What travis can't. 

Doesn't qemu-user binfmt work on Travis?

Reply:
Emulation is not that stable that we could trust it in here. Especially on power/s390x. Also, I'd expect this script to run anywhere, without requiring special setup or a new kernel.

Reply:
(haven't really tested if emultaion works at all on travis)

Reply:
Could you add the following stuff to the comment lines of the Dockerfile:
* image digest for reproducibility
* how you built the image
* background (can be just a link to this GitHub issue thread)